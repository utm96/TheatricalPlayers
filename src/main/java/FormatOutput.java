public enum FormatOutput {
    HTML, TEXT
}
