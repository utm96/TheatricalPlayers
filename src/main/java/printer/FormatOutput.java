package printer;

public enum FormatOutput {
    HTML,TEXT
}
