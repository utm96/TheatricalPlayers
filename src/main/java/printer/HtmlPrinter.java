package printer;

import result.Line;
import result.StatementResult;

import java.text.NumberFormat;
import java.util.List;
import java.util.Locale;

public class HtmlPrinter implements Printer {
    @Override
    public String print(StatementResult statementResult) {
        NumberFormat numberFormat = NumberFormat.getCurrencyInstance(Locale.US);
        StringBuilder result = new StringBuilder(printCustomer(statementResult));
        List<Line> lines = statementResult.getLines();
        for (Line line : lines) {
            result.append(printLine(line, numberFormat));
        }
        result.append(printTotalAmount(statementResult, numberFormat));
        result.append(printVolumeCredit(statementResult));
        result.append(endOfFile());
        return result.toString();
    }

    private String printCustomer(StatementResult statementResult) {
        return String.format("""
                <!DOCTYPE html>
                <html>
                <head>
                    <style>
                table, th, td {
                border: 1px solid black;
                }
                th, td {
                padding: 5px;
                }
                </style>
                </head>
                <body>
                <h1>Invoice</h1>
                <p><strong>Client:</strong> %s</p>
                <table>
                    <tr>
                        <th>Piece</th>
                        <th>Seats Sold</th>
                        <th>Price</th>
                    </tr>
                """, statementResult.getCustomer());
    }

    private String printLine(Line line, NumberFormat frmt) {
        return String.format("""
                    <tr>
                        <td>%s</td>
                        <td>%s</td>
                        <td>%s</td>
                    </tr>
                """, line.getName(), line.getAudience(), frmt.format(line.getAmount() / 100));
    }

    private String printTotalAmount(StatementResult statementResult, NumberFormat frmt) {
        return String.format("""
                    <tr>
                        <td colspan="2" style="text-align: right;"><strong>Total Owed:</strong></td>
                        <td>%s</td>
                    </tr>
                """, frmt.format(statementResult.getTotalAmount() / 100));
    }

    private String printVolumeCredit(StatementResult statementResult) {
        return String.format("""
                    <tr>
                        <td colspan="2" style="text-align: right;"><strong>Fidelity Points Earned:</strong></td>
                        <td>%s</td>
                    </tr>
                """, statementResult.getVolumeCredit());
    }
    private String endOfFile() {
        return """
                </table>

                <p><strong>Payment is required under 30 day. We can break your knees if you don't do so</strong></p>
                </body>
                </html>""";
    }
}
