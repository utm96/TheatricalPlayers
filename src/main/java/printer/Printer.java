package printer;

import result.StatementResult;

public interface Printer {
    String print(StatementResult statementResult);
}
