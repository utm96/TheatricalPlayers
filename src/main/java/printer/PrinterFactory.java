package printer;

public class PrinterFactory {
    public static Printer printerFor(FormatOutput formatOutput) {
        switch (formatOutput) {
            case HTML -> {
                return new HtmlPrinter();
            }
            default -> {
                return new TextPrinter();
            }
        }
    }
}
