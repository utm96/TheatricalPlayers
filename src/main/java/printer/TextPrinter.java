package printer;

import result.Line;
import result.StatementResult;

import java.text.NumberFormat;
import java.util.List;
import java.util.Locale;
import java.util.Optional;

public class TextPrinter implements Printer{
    @Override
    public String print(StatementResult statementResult) {
        NumberFormat frmt = NumberFormat.getCurrencyInstance(Locale.US);
        StringBuilder result = new StringBuilder(String.format("Statement for %s\n", statementResult.getCustomer()));
        List<Line> lines = statementResult.getLines();
        for (Line line : lines) {
            result.append(String.format("  %s: %s (%s seats)\n", line.getName(), frmt.format(line.getAmount() / 100), line.getAudience()));
        }
        result.append(String.format("Amount owed is %s\n", frmt.format(statementResult.getTotalAmount() / 100)));
        result.append(String.format("You earned %s credits\n", statementResult.getVolumeCredit()));
        return result.toString();
    }

}
