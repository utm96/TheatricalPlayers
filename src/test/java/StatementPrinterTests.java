import org.approvaltests.core.Options;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import printer.FormatOutput;

import java.util.List;
import java.util.Map;

import static org.approvaltests.Approvals.verify;
import static org.approvaltests.Approvals.verifyHtml;

public class StatementPrinterTests {

    @Test
    void exampleStatement() {
        Map<String, Play> plays = Map.of(
                "hamlet",  new Play("Hamlet", "tragedy"),
                "as-like", new Play("As You Like It", "comedy"),
                "othello", new Play("Othello", "tragedy"));

        Invoice invoice = new Invoice("BigCo", List.of(
                new Performance("hamlet", 55),
                new Performance("as-like", 35),
                new Performance("othello", 40)));

        StatementPrinter statementPrinter = new StatementPrinter();
        var result = statementPrinter.print(invoice, plays, FormatOutput.TEXT);

        verify(result);
    }


    @Test
    void exampleStatementHtml() {
        Map<String, Play> plays = Map.of(
                "hamlet",  new Play("Hamlet", "tragedy"),
                "as-like", new Play("As You Like It", "comedy"),
                "othello", new Play("Othello", "tragedy"));

        Invoice invoice = new Invoice("BigCo", List.of(
                new Performance("hamlet", 55),
                new Performance("as-like", 35),
                new Performance("othello", 40)));

        StatementPrinter statementPrinter = new StatementPrinter();
        var result = statementPrinter.print(invoice, plays,FormatOutput.HTML);
//        System.out.println(result);
        verifyHtml(result);
    }

    @Test
    void statementWithNewPlayTypes() {
        Map<String, Play> plays = Map.of(
                "henry-v",  new Play("Henry V", "history"),
                "as-like", new Play("As You Like It", "pastoral"));

        Invoice invoice = new Invoice("BigCo", List.of(
                new Performance("henry-v", 53),
                new Performance("as-like", 55)));

        StatementPrinter statementPrinter = new StatementPrinter();
        Assertions.assertThrows(Error.class, () -> {
            statementPrinter.print(invoice, plays,FormatOutput.TEXT);
        });
    }
}
