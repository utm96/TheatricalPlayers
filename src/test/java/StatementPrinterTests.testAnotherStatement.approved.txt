Statement for BigCo
  Hamlet: $400.00 (30 seats)
  As You Like It: $360.00 (20 seats)
Amount owed is $760.00
You earned 4 credits
